main.c

```diff
@@ -0,0 +1,315 @@
+struct tee_ts_global *bench_ts_global;
+static bool is_running;
+
+static TEEC_SharedMemory ts_buf_shm = {
+		.flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT,
+		.buffer = NULL,
+		.size = 0
+};
+
+static TEEC_Context ctx;
+static TEEC_Session sess;
+
+static void open_bench_pta(void)
+{
+	TEEC_Result res;
+	uint32_t err_origin;
+
+	res = TEEC_InitializeContext(NULL, &ctx);
+	tee_check_res(res, "TEEC_InitializeContext");
+
+	res = TEEC_OpenSession(&ctx, &sess, &pta_benchmark_uuid,
+			TEEC_LOGIN_PUBLIC, NULL, NULL, &err_origin);
+	tee_check_res(res, "TEEC_OpenSession");
+}
+
+static void close_bench_pta(void)
+{
+	/* release benchmark timestamp shm */
+	TEEC_ReleaseSharedMemory(&ts_buf_shm);
+	TEEC_CloseSession(&sess);
+	TEEC_FinalizeContext(&ctx);
+}
+
+static void init_ts_global(struct tee_ts_global *ts_global,
+				uint32_t cores)
+{
+	struct tee_ts_cpu_buf *cpu_buf;
+
+	/* init global timestamp buffer */
+	bench_ts_global = ts_global;
+	bench_ts_global->cores = cores;
+
+	/* init per-cpu timestamp buffers */
+	for (int i = 0; i < cores; i++) {
+		cpu_buf = &bench_ts_global->cpu_buf[i];
+		memset(cpu_buf, 0, sizeof(struct tee_ts_cpu_buf));
+	}
+}
+
+static void register_bench_buf(uint32_t cores)
+{
+	TEEC_Result res;
+	TEEC_Operation op = { 0 };
+	uint32_t ret_orig;
+
+	ts_buf_shm.size = sizeof(struct tee_ts_global)
+			+ sizeof(struct tee_ts_cpu_buf) * cores;
+
+	/* allocate global timestamp buffer */
+	res = TEEC_AllocateSharedMemory(&ctx, &ts_buf_shm);
+	tee_check_res(res, "TEEC_AllocateSharedMemory");
+
+	init_ts_global((struct tee_ts_global *)ts_buf_shm.buffer, cores);
+
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_PARTIAL_INOUT,
+			TEEC_NONE, TEEC_NONE, TEEC_NONE);
+	op.params[0].memref.parent = &ts_buf_shm;
+	op.params[0].memref.offset = 0;
+
+	TEEC_InvokeCommand(&sess, BENCHMARK_CMD_REGISTER_MEMREF,
+					&op, &ret_orig);
+	tee_check_res(res, "TEEC_InvokeCommand");
+}
+
+static void unregister_bench(void)
+{
+	TEEC_Result res;
+	TEEC_Operation op = { 0 };
+	uint32_t ret_orig;
+
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_NONE,
+					TEEC_NONE, TEEC_NONE, TEEC_NONE);
+
+	res = TEEC_InvokeCommand(&sess, BENCHMARK_CMD_UNREGISTER,
+					&op, &ret_orig);
+	tee_check_res(res, "TEEC_InvokeCommand");
+}
+
+static void usage(char *progname)
+{
+	fprintf(stderr, "Call latency benchmark tool for OP-TEE\n\n");
+	fprintf(stderr, "Usage:\n");
+	fprintf(stderr, "  %s -h\n", progname);
+	fprintf(stderr, "  %s host_app [host_app_args]\n", progname);
+	fprintf(stderr, "Options:\n");
+	fprintf(stderr, "  -h              Print this help and exit\n");
+	fprintf(stderr, "  host_app        Path to host app to benchmark\n");
+	fprintf(stderr, "  host_app_args   Original host app args\n");
+}
+
+static int timestamp_pop(struct tee_ts_cpu_buf *cpu_buf,
+						struct tee_time_st *ts)
+{
+	struct tee_time_st ts_data;
+	uint64_t ts_tail = 0;
+
+	if (!cpu_buf && !ts)
+		return RING_BADPARM;
+
+	if (cpu_buf->tail >= cpu_buf->head)
+		return RING_NODATA;
```

Comment:
both `head` and `tail ` are `uint64_t`, and they are not actually an index in the circular buffer. Index is  `ts_i & TEE_BENCH_MAX_MASK`

 > `tail` might get higher than `head` and buffer can still hold valid content.

this could be true only if `head` overflows, which is almost an impossible case, if to consider an intensity of adding timestamps